doc/Networking_Userland.md

```diff
@@ -0,0 +1,266 @@
+        Returns EINVAL if no valid buffer has been loaded into the write buffer,
+        or if the config buffer is the wrong length, or if the destination and source
+        oirt/address pairs cannot be parsed.
+        Otherwise, returns the result of do_next_tx_sync(). Notably, a successful
+        transmit can produce two different success values. If success is returned,
+        this simply means that the packet was queued. However, if SuccessWithValue
+        is returned with value 1, this means the the packet was successfully passed
+        the radio without any errors, which tells the userland application that it can
+        immediately queue another packet without having to wait for a callback.
+
+udp.c and udp.h in libtock-c make it easy to interact with this driver interface.
+Important functions available to userland apps written in c include:
+
+`udp_socket()` - sets the port on which the app will receive udp packets,
+                 and sets the `src_port` of outgoing packets sent via that socket. Once socket
+                 binding is implemented in the kernel, this function will handle reserving ports
+                 to listen on and send from.
+
+`udp_close()` - currently just returns success, but once socket binding has been
+                implemented in the kernel, this function will handle freeing bound ports.
+
+`udp_send_to()` - Sends a udp packet to a specified addr/port pair, returns the result
+                  of the tranmission once the radio has transmitted it (or once a failure has occured).
+
+`udp_recv_from_sync()` - Pass an interface to listen on and an incoming source address
+                         to listen for. Sets up a callback to wait for a received packet, and yeilds until that
+                         callback is triggered. This function never returns if a packet is not received.
+
+`udp_recv_from()` - Pass an interface to listen on and an incoming source address to
+                    listen for. However, this takes in a buffer to which the received packet should be placed,
+                    and returns the callback that will be triggered when a packet is received.
+
+`udp_list_ifaces()` - Populates the passed pointer of ipv6 addresses with the available
+                      ipv6 addresses of the interfaces on the device. Right now this merely returns a constant
+                      hardcoded into the UDP driver, but should change to return the source IP addresses held
+                      in the network configuration file once that is created. Returns up to `len` addresses.
+
+Other design notes:
+
+The current design of the driver has a few limitations, these include:
+
+- Currently, any app can listen on any address/port pair
+
+- The current tx implementation allows for starvation, e.g. an app with an earlier app ID can
+  starve a later ID by sending constantly.
+
+### POSIX Socket API Functions
```

Comment:
I'm not sure we need this, but if it is useful can it be added to the networking stack document? Having a userland interface is a part of the full networking stack.